version/info_go118.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2022 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package version
+
+import "runtime/debug"
+
+var computedRevision string
+
+func getRevision() string {
+	return computedRevision
+}
+
+func init() {
+	computedRevision = computeRevision()
+}
+
+func computeRevision() string {
+	var (
+		rev      = "unknown"
+		modified bool
+	)
+
+	buildInfo, ok := debug.ReadBuildInfo()
+	if !ok {
+		return rev
+	}
+	for _, v := range buildInfo.Settings {
+		if v.Key == "vcs.revision" {
+			rev = v.Value
+		}
+		if v.Key == "vcs.modified" {
+			if v.Value == "true" {
+				modified = true
+			}
+		}
+	}
+	if modified {
+		return rev + "-modified"
+	}
```

Comment:
I've seen this returned as a separate parameter, this is basically checking if `git` was in a dirty state at the time of build.

Reply:
changing the API is out of scope here I'd say, as we see revision as one. it should still be pretty easy to parse as modified is not a valid git hash